config_parser.cc

```diff
@@ -198,15 +200,21 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
       config_stack.top()->statements_.back().get()->child_block_.reset(
           new_config);
       config_stack.push(new_config);
+      // Increment unmatchedBrackets to make sure there's closing bracket later
+      unmatchedBrackets++;
     } else if (token_type == TOKEN_TYPE_END_BLOCK) {
-      if (last_token_type != TOKEN_TYPE_STATEMENT_END) {
+      // Only tokens before this can be ; or }
+      if (!(last_token_type == TOKEN_TYPE_STATEMENT_END
+        || last_token_type == TOKEN_TYPE_END_BLOCK)) {
         // Error.
         break;
       }
       config_stack.pop();
+      unmatchedBrackets--;
     } else if (token_type == TOKEN_TYPE_EOF) {
-      if (last_token_type != TOKEN_TYPE_STATEMENT_END &&
-          last_token_type != TOKEN_TYPE_END_BLOCK) {
+      if ((last_token_type != TOKEN_TYPE_STATEMENT_END &&
+          last_token_type != TOKEN_TYPE_END_BLOCK)
+        || unmatchedBrackets != 0) { // Number of starting brackets must match closing brackets
```

Comment:
Consider creating a separate if statement in order to provide your own error message.